kong/plugins/ai-response-transformer/handler.lua

```diff
@@ -0,0 +1,163 @@
+local _M = {}
+
+-- imports
+local kong_meta     = require "kong.meta"
+local http          = require("resty.http")
+local fmt           = string.format
+local kong_utils    = require("kong.tools.utils")
+local llm           = require("kong.llm")
+--
+
+_M.PRIORITY = 769
+_M.VERSION = kong_meta.version
+
+local function bad_request(msg)
+  kong.log.warn(msg)
+  return kong.response.exit(400, { error = { message = msg } })
+end
+
+local function internal_server_error(msg)
+  kong.log.err(msg)
+  return kong.response.exit(500, { error = { message = msg } })
+end
+
+local function subrequest(httpc, request_body, http_opts)
+  httpc:set_timeouts(http_opts.http_timeout or 60000)
+
+  local upstream_uri = ngx.var.upstream_uri
+  if ngx.var.is_args == "?" or string.sub(ngx.var.request_uri, -1) == "?" then
+    ngx.var.upstream_uri = upstream_uri .. "?" .. (ngx.var.args or "")
+  end
+
+  local ok, err = httpc:connect {
+    scheme = ngx.var.upstream_scheme,
+    host = ngx.ctx.balancer_data.host,
+    port = ngx.ctx.balancer_data.port,
+    proxy_opts = http_opts.proxy_opts,
+    ssl_verify = http_opts.ssl_verify or true,
+    ssl_server_name = ngx.ctx.balancer_data.host,
+  }
+
+  if not ok then
+    return nil, internal_server_error("failed to connect to upstream: " .. err)
```

Comment:
This seems to be redundant with the error-handling in `_M:access()`:

```lua
  local res, err = subrequest(httpc, kong.request.get_raw_body(), http_opts)
  if err then
    return internal_server_error(err)
  end
```

So I would recommend returning an error string instead:

```suggestion
    return nil, "failed to connect to upstream: " .. err
```